Mixnet integration

zdave opened the pull request:
See paritytech/substrate#13904.

This adds all the necessary mixnet components, and puts them together in the "kitchen-sink" node/runtime. The components added are:

- A pallet (`frame/mixnet`). This is responsible for determining the current mixnet session and phase, and the mixnodes to use in each session. It includes an offchain worker which attempts to automatically register validator nodes as mixnodes. The logic of this pallet is very similar to that of the `im-online` pallet.
- A service (`client/mixnet`). This implements the core mixnet logic, building on the [mixnet crate](https://crates.io/crates/mixnet). The service communicates with other nodes using notifications sent over the "mixnet" protocol.
- An RPC interface. This currently only supports sending transactions over the mixnet.

zdave committed:
Disconnect peer on unreserving if no free slot for

zdave committed:
Mixnet pallet

Responsible for determining the mixnode set for each session.

zdave committed:
Mixnet service

zdave committed:
Mixnet RPC interface

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2883299

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2886471

zdave commented:
Ideally the new CLI options would only be added to the kitchen-sink node, but the best way to achieve this is not obvious to me.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2886668

cheme reviewed (commented):
Had a few question from my first reading, at this point I did not yet get the session even odd logic, but will look at it soon.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,762 @@
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Register a mixnode for the following session.
```
No. Doesn't feel worthwhile to implement this to me; sessions should only be a few hours?

cheme replied:
if a few hour offline is acceptable (probably is initially).

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,762 @@
+
+			// Check signature
+			let signature_ok = registration.using_encoded(|encoded_registration| {
+				authority_id.verify(&encoded_registration, signature)
```
Unsigned transactions are used so that validators don't have to pay a fee to register. To avoid spam, each validator can only register once per session. This is very similar to how the `im-online` pallet works.

cheme replied:
I don't know if `im-online` work this way I can probably same im-online transaction with no consequences.  Can say it is a bit out of my expertise area though.

zdave commented on client/mixnet/src/request.rs:
```diff
@@ -0,0 +1,109 @@
+
+pub fn extrinsic_delay(message_id: &MessageId, config: &SubstrateConfig) -> Duration {
+	let h = Blake2bMac::<U16>::new_with_salt_and_personal(message_id, b"", EXTRINSIC_DELAY_PERSONA)
+		.expect("Key, salt, and persona sizes are fixed and small enough");
```
Both the sender and receiver calculate this; the sender uses it to determine the timeout. Only the sender and receiver should know the message ID, so I think this is fine. An alternative would be to just randomly pick a delay in the receiver and have the sender use the 99.9th percentile for the timeout. I think having the sender know the exact delay is preferable as it allows for tighter timeouts, but I could probably be convinced otherwise!

cheme replied:
Cannot really say, but sure is something to keep in mind (should be in the top level documentation).
Is the message ID the same for every hops?

tomaka commented:
Is there (or is there going to be) some kind of documentation for how this works at the protocol level somewhere?

cheme commented:
> Is there (or is there going to be) some kind of documentation for how this works at the protocol level somewhere? (source code doesn't count, since it is extremely hard to decipher for non-maintainers)

That is a really good question, the design did change a lot since the first prototyping, and a high level documentation would be probably good too.
Not sure when is the right time for the effort, and the correct medium.
I mean probably what I would want is a good technical blog post to give me reading direction on the project (with frame description and more general network behaviour (delays, routings...)).
Could also wonder if the protocol could be in a Psp?

arkpar commented:
Here's a tracking issue for the documentaiton:
https://github.com/paritytech/mixnet/issues/13

zdave commented:
> Is there (or is there going to be) some kind of documentation for how this works at the protocol level somewhere? (source code doesn't count, since it is extremely hard to decipher for non-maintainers)

There isn't really anything yet. I'll try to put something together next week. If you're interested in something to read now, the packet format is very close to what is described in the [Sphinx paper](https://www.researchgate.net/publication/220713667_Sphinx_A_Compact_and_Provably_Secure_Mix_Format).

zdave commented on client/api/src/execution_extensions.rs:
```diff
@@ -165,6 +168,7 @@ impl<T: offchain::DbExternalities + Clone + Sync + Send + 'static> DbExternaliti
 pub struct ExecutionExtensions<Block: BlockT> {
 	strategies: ExecutionStrategies,
 	keystore: Option<KeystorePtr>,
+	mixnet_kx_public_store: Option<Arc<MixnetKxPublicStore>>,
```
I didn't think the requirements overlapped enough for this to make sense. The existing keystore is (AFAICT) designed for long-lived keys that persist across node restarts. The mixnet key-exchange keys are only used for a single session (a few hours), and should be destroyed once the session ends. They intentionally do not persist across node restarts, as this would be non-trivial to implement properly:

- When destroying keys you would have to be careful to properly erase the key data from disk.
- The corresponding replay filters would also have to persist, or you would be vulnerable to replay attacks after a node restart.

I opened a GitHub issue about this a while back (#12595), there wasn't really any conclusion though.

zdave commented on client/cli/src/params/mixnet_params.rs:
```diff
@@ -0,0 +1,56 @@
+	/// The mixnet key-exchange secret to use in session 0.
+	///
+	/// Should be 64 hex characters, giving a 32-byte secret. This option should be used for
+	/// testing purposes only.
```
Sure, will do.

arkpar replied:
As far as I understand this is only used for session 0. Any real chain will have an existing session and will use an on-chain key anyway.

zdave commented on primitives/mixnet/runtime-api/Cargo.toml:
```diff
@@ -0,0 +1,26 @@
+[package]
+description = "Substrate mixnet runtime API"
+name = "sp-mixnet-runtime-api"
```
AFAICT all runtime APIs are in `primitives`.

zdave commented on client/mixnet/src/request.rs:
```diff
@@ -0,0 +1,109 @@
+
+	fn num_surbs(&self, context: &Self::Context) -> usize {
+		match self {
+			Request::SubmitExtrinsic { .. } => context.surb_factor,
```
It's not really possible to disable replies. If you set `surb_factor` to 0 this _would_ prevent any replies from being sent, but the sender would still expect one, so it wouldn't work very well!

zdave replied:
I've added a little note to the `surb_factor` doc comment that it shouldn't be set to 0.

arkpar commented on client/mixnet/src/packet_dispatcher.rs:
```diff
@@ -0,0 +1,208 @@
+	/// Push `packet` onto the queue. Returns `true` if the queue was previously empty. Fails if
+	/// the queue is full.
+	fn push(&self, packet: Box<Packet>) -> Result<bool, ()> {
+		let mut queue = self.0.lock().unwrap();
```
Consider using `parking_lot::Mutex`. It's what we use all over substrate and it has infallible lock.

zdave replied:
Switched to `parking_lot::Mutex`.

arkpar commented on client/mixnet/src/request.rs:
```diff
@@ -0,0 +1,109 @@
+	}
+}
+
+pub const SUBMIT_EXTRINSIC: u8 = 0;
```
I'd start with 1 so that 0 is reserved and can be used to detect invalid/corrupted data more reliably.

arkpar commented:
From what I can see, after the initial sync the node will wait for the registration phase of the next session before registering. How does it it handle traffic in the meantime and what key will be used? Would it make sense to register immediatelly after the sync is completed?

arkpar commented on primitives/io/src/lib.rs:
```diff
@@ -1443,6 +1447,19 @@ pub trait Offchain {
 
+/// Interface for querying the local node's mixnet keystore.
+#[runtime_interface]
+pub trait MixnetKxPublicStore {
```
This could be moved to e.g.`primitives/mixnet/runtime-api`.

zdave replied:
AFAICT all the other runtime->host interfaces are defined in this file? `primitives/mixnet/runtime-api` has the host->runtime interface.

zdave replied:
I see what you mean now. Done this.

zdave commented:
> From what I can see, after the initial sync the node will wait for the registration phase of the next session before registering. How does it it handle traffic in the meantime and what key will be used? Would it make sense to register immediatelly after the sync is completed?

If a node isn't registered as a mixnode it should only receive reply and loop cover traffic, which it will generate a random key pair for. The public key doesn't need to be shared with any other nodes.

If a node is registered as a mixnode and then restarts, it will no longer have the private key corresponding to the public key that was registered, so it will only be able to operate as a non-mixnode until the next session.

I haven't really thought about allowing registrations after the start of a session. Maybe this could be made to work, but it seems like a bunch of extra complexity for little gain. Sessions are only a few hours long, and if a node is an active validator in a session, it should really stay online for the whole session.

zdave commented on client/mixnet/src/extrinsic_queue.rs:
```diff
@@ -0,0 +1,88 @@
+}
+
+impl<E> ExtrinsicQueue<E> {
+	pub fn new(capacity: usize) -> Self {
```
This isn't visible outside of the crate. I'll add more comments if CI complains.

zdave commented on client/mixnet/src/sync_with_runtime.rs:
```diff
@@ -0,0 +1,134 @@
+	B: Block,
+	A: MixnetApi<B>,
+{
+	let session_status = match api.session_status(hash) {
```
`session_status` doesn't just return the session index, it also returns the phase within the session. I put all the phase determination logic in the pallet as everyone needs to agree on it, even though the pallet itself only really cares about the registration phase.

Mixnet sessions do not really need to match up with BABE sessions or whatever, that's kind of an implementation detail of the pallet. `session_status` need not even be backed by the mixnet pallet, although of course it is expected to be. For these reasons I don't think we should talk about how `session_status` is implemented here.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,762 @@
+
+impl<MaxPeerIdSize, MaxExternalAddressSize, MaxExternalAddresses> Into<OpaqueMixnode>
+	for BoundedOpaqueMixnode<
+		BoundedVec<u8, MaxPeerIdSize>,
```
I would like to write something like `impl<T: Config> Into<OpaqueMixnode> for BoundedOpaqueMixnodeFor<T>` here, but the Rust compiler doesn't like this:

    error[E0207]: the type parameter `T` is not constrained by the impl trait, self type, or predicates
      --> frame/mixnet/src/lib.rs:84:6
       |
    84 | impl<T: Config> Into<OpaqueMixnode> for BoundedOpaqueMixnodeFor<T>
       |      ^ unconstrained type parameter

I can do something like this, but not sure it's better:

    pub type BoundedPeerId<MaxSize> = BoundedVec<u8, MaxSize>;
    pub type BoundedExternalAddress<MaxSize> = BoundedVec<u8, MaxSize>;
    pub type BoundedExternalAddresses<MaxSizeEach, Max> =
    	BoundedVec<BoundedExternalAddress<MaxSizeEach>, Max>;
    
    impl<MaxPeerIdSize, MaxExternalAddressSize, MaxExternalAddresses> Into<OpaqueMixnode>
    	for BoundedOpaqueMixnode<
    		BoundedPeerId<MaxPeerIdSize>,
    		BoundedExternalAddresses<MaxExternalAddressSize, MaxExternalAddresses>,
    	>

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2929338

zdave commented on client/api/src/execution_extensions.rs:
```diff
@@ -244,6 +250,12 @@ impl<Block: BlockT> ExecutionExtensions<Block> {
 		}
 
+		if capabilities.contains(offchain::Capabilities::MIXNET_KX_PUBLIC_STORE) {
+			if let Some(ref mixnet_kx_public_store) = self.mixnet_kx_public_store {
```
Done.

zdave commented on client/mixnet/src/sync_with_runtime.rs:
```diff
@@ -0,0 +1,134 @@
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
```
I've added a few module doc comments in `client/mixnet`. I haven't put anything here about what happens on session change, as that's not really the responsibility of this code; it merely sets the session status in the core mixnet state.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,762 @@
+		}
+
+		// Equivalent; ok to replace if we have waited long enough
+		self.block_number.saturating_add(3_u32.into()) < other.block_number
```
Done.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,762 @@
+
+	fn mixnodes(current: bool) -> Result<Vec<OpaqueMixnode>, MixnodesErr> {
+		let mixnodes: Vec<_> =
+			if (T::ValidatorSet::session_index() & 1) == (current as SessionIndex) {
```
Hopefully this is all a bit clearer now. I've added a doc comment to this function to explain what `current` means, and added a `with_mixnodes` macro to hide the `OddSessionMixnode`/`EvenSessionMixnodes` logic.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2929660

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2930065

zdave-parity committed:
Basic mixnet CLI options

zdave-parity committed:
Add mixnet to kitchen sink node/runtime

zdave commented:
> Ideally the new CLI options would only be added to the kitchen-sink node, but the best way to achieve this is not obvious to me.

Done this now. And got rid of the `client/service` changes; all that stuff is just in `bin/node` now.

zdave commented:
> > Is there (or is there going to be) some kind of documentation for how this works at the protocol level somewhere? (source code doesn't count, since it is extremely hard to decipher for non-maintainers)
> 
> There isn't really anything yet. I'll try to put something together next week. If you're interested in something to read now, the packet format is very close to what is described in the [Sphinx paper](https://www.researchgate.net/publication/220713667_Sphinx_A_Compact_and_Provably_Secure_Mix_Format).

I've put something together here: https://github.com/zdave-parity/mixnet-spec. Feedback welcome!

arkpar commented:
@bkchr Please review or delegate

bkchr reviewed (commented):
The general inclusion of the mixnet in the node looks good to me :+1: 

However, I have left some questions to the design of the mixnet implementation in Substrate. The two main questions I left are around the offchain worker and the storage of the peer information.

1. I think we should replace the offchain worker just with a runtime api function that then constructs the extrinsic and sends it to the tx pool. The advantage of this would be that we would not require a new  host function as we could pass the `kx_public` as parameter to this runtime api function.
2. I would not store the `PeerId` & `MultiAddr` on chain. We already have this information available in the authority disocvery, so there is no need to keep it in the state. In general someone could also argue this about `kx_public`. It is not really required in the state and the state is just "abused" as a storage. This key could also go in the DHT or be used as session key directly (but that would require for changes, which are also questionable, see one of my comments).

I assume the requirement to change `kx_public` every session is to make it harder to "decrypt" the traffic?

bkchr commented:
And another question, what happens if a node doesn't participate as mix node (while it is expected from the on chain state)? Currently we assume that all validators are taking part in the protocol. However, there is also no real incentivization for them to take part in the protocol. It will also constantly generate traffic which may let's certain people try to disable it.

zdave commented:
> 1. I think we should replace the offchain worker just with a runtime api function that then constructs the extrinsic and sends it to the tx pool. The advantage of this would be that we would not require a new  host function as we could pass the `kx_public` as parameter to this runtime api function.

Sure, I can do this. The main reason it currently uses an offchain worker is because that's what the im-online pallet uses, which I cribbed from.

> 2. I would not store the `PeerId` & `MultiAddr` on chain. We already have this information available in the authority disocvery, so there is no need to keep it in the state. In general someone could also argue this about `kx_public`. It is not really required in the state and the state is just "abused" as a storage. This key could also go in the DHT or be used as session key directly (but that would require for changes, which are also questionable, see one of my comments).

I was originally using the authority discovery module, but changed this after discussion with Arkadiy and Emeric. The main reasons for preferring to store `PeerId` and `Multiaddr` on chain are:

- It will make transitioning away from mixnodes-are-always-authorities easier.
- It's much simpler.

Possibly there were other reasons that came up, I don't recall the details of the discussion any more.

As for storing `kx_public` in the DHT this just seems like a bunch more complexity for little gain. Accessing these keys over the DHT could also open up new avenues for attack; what do you do if you cannot get hold of all of the public keys?

> I assume the requirement to change `kx_public` every session is to make it harder to "decrypt" the traffic?

It's to provide forward secrecy. The secret keys for a session are erased after the session ends, so if a node's keys get leaked somehow, the impact is limited; old traffic cannot be decrypted.

zdave commented:
> It would also be nice to have some Zombienet test that ensures that sending a transaction via mixnet works as expected.

Yes, I do plan to add some Zombienet tests. There is a bug that needs fixing in Zombienet first... see [here](https://github.com/paritytech/zombienet/issues/1155).

cheme commented:
> > It would also be nice to have some Zombienet test that ensures that sending a transaction via mixnet works as expected.
> 
> Yes, I do plan to add some Zombienet tests. There is a bug that needs fixing in Zombienet first... see [here](https://github.com/paritytech/zombienet/issues/1155).

I did have this one at the time, but probably zombie net did change a lot from this date https://github.com/cheme/substrate/tree/mixnet_2/zumbanet_tests/mixnet .

zdave commented:
> And another question, what happens if a node doesn't participate as mix node (while it is expected from the on chain state)? Currently we assume that all validators are taking part in the protocol. However, there is also no real incentivization for them to take part in the protocol. It will also constantly generate traffic which may let's certain people try to disable it.

Incentivisation isn't something that I've tried to address in this first version. I believe @burdges has a scheme, but I don't know the details.

If someone running a validator disables the mixnet module, then the node will not try to register as a mixnode. So we aren't really assuming all validators will act as mixnodes, it's more that we don't allow non-validators to act as mixnodes. If too few validators register as mixnodes then the mixnet will just not run.

Obviously there's nothing to stop someone registering as a mixnode and then not participating. This will cause packets sent via the node to get lost. If enough nodes do this then the mixnet will essentially become unusable.

zdave commented on client/cli/src/params/mixnet_params.rs:
```diff
@@ -0,0 +1,66 @@
+				},
+				..Default::default()
+			};
+			// Only mixnodes, which are always authorities, need to allow connections from
+			// non-mixnodes
```
I guess this could be worded better. By "mixnodes are always authorities" I mean only the authorities in a session can be mixnodes; non-authorities cannot register as mixnodes. Long term I expect this to change.

zdave replied:
This has been reworded, hopefully clear now.

zdave commented on client/mixnet/src/config.rs:
```diff
@@ -0,0 +1,90 @@
+
+use std::time::Duration;
+
+/// Substrate-specific mixnet configuration.
+#[derive(Clone, Debug)]
+pub struct SubstrateConfig {
```
I agree that the name is not great, but I don't have any better ideas. I named this `SubstrateConfig` because the other parts of the configuration are defined in the `mixnet` crate and are thus not specific to Substrate.

zdave replied:
Think I'll keep it as-is. The other config structs are handed off to the relevant subsystems in `run`. References to the `SubstrateConfig` are passed around within `sc-mixnet`. Could pass around `Config` references instead, but that would require extra cloning and pointlessly give things access to `CoreConfig` etc.

zdave commented on client/mixnet/src/extrinsic_queue.rs:
```diff
@@ -0,0 +1,94 @@
+use mixnet::reply_manager::ReplyContext;
+use std::{cmp::Ordering, collections::BinaryHeap, time::Instant};
+
+/// `Eq` and `Ord` are implemented for this to support use in `BinaryHeap`s. Only `deadline` is
+/// compared.
```
This is a private struct so not sure it really matters, but I don't agree with "this is a code comment as doc". Happy to put the suggested doc comment first in any case.

zdave commented on primitives/core/src/offchain/mod.rs:
```diff
@@ -280,6 +280,8 @@ bitflags::bitflags! {
 		const TIME = 0b0001_0000_0000;
 		/// Access the statement store.
 		const STATEMENT_STORE = 0b0010_0000_0000;
+		/// Access to local node's mixnet key-exchange public keys.
+		const MIXNET_KX_PUBLIC_STORE = 0b0100_0000_0000;
```
It is used in `execution_extensions.rs`. However I'm happy to switch away from the offchain worker as suggested, in which case this will not be needed.

zdave replied:
Done.

zdave commented on client/rpc/src/mixnet/mod.rs:
```diff
@@ -0,0 +1,47 @@
+			let mut api = self.0.lock().await;
+			api.submit_extrinsic(extrinsic).await
+		};
+		Ok(fut.await.map_err(Error)?)
```
This won't compile; `?` is implicitly calling `Into::<jsonrpsee::core::Error>::into` on the `Error`.

bkchr commented:
> * It will make transitioning away from mixnodes-are-always-authorities easier.

Everyone can put data into the DHT, not only authorities. It is just that we currently do it this way. IMO it would probably also be much easier to use the DHT for non-authority nodes in the future. Especially with something like this: https://github.com/paritytech/substrate/issues/13248 where you could then "search" for nodes participating in mixnet.



> Possibly there were other reasons that came up, I don't recall the details of the discussion any more

@cheme any idea?



> Incentivisation isn't something that I've tried to address in this first version. I believe @burdges has a scheme, but I don't know the details.

Yeah, this is just that came up in my mind while thinking about this pr. I don't want to have a solution for this right now :D

zdave commented on client/mixnet/src/packet_dispatcher.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+/// A peer which has packets ready to send but is not currently being serviced.
+pub struct ReadyPeer {
+	id: PeerId,
+	/// The peer's packet queue. Not empty.
+	queue: Arc<PeerQueue>,
```
A `ReadyPeer` doesn't live very long, but the `PeerQueue` reference does; it is stored in the future returned by `ReadyPeer::send_packet`. This future is stuffed in a `FuturesUnordered` and may still be alive when `PacketDispatcher::dispatch` is called again. `dispatch` may push more packets into the queue. A simple borrow of the queue is not sufficient.

As the comment above `PeerQueue` says, `Arc<Mutex<_>>` isn't really needed, as everything is effectively single-threaded; `Rc<RefCell<_>>` would do. I don't think it would be possible to have the `run` future be `Send` using `Rc<RefCell<_>>` though, without using `unsafe`. `Arc<Mutex<_>>` isn't that expensive when there is no contention, so I wasn't particularly worried about this.

bkchr replied:
Ahh, I had `Ok(more_packets) => more_packets.then(|| self)` overseen. Yeah, then my comment is useless^^

zdave commented on client/mixnet/src/request.rs:
```diff
@@ -0,0 +1,112 @@
+	}
+}
+
+pub const SUBMIT_EXTRINSIC: u8 = 1;
```
Mainly because it's simpler and more efficient to just handle the top-level enum manually. Note that the encoding _is_ effectively SCALE, the library just isn't used for handling the top-level enum.

bkchr replied:
I'm not really sure we need this small benefit of being more performant vs better readability.

zdave replied:
I don't agree about the readability honestly. It's straight forward as-is: the client side sticks a `SUBMIT_EXTRINSIC` byte at the start and the server side does a match on the first byte. In my eyes that is simpler and easier to understand than encode/decode calls.

Note that with the `Request` enum as you have defined it you are going to effectively double-encode the extrinsic, which is a bit confusing IMO.

cheme commented:
> @cheme any idea?

I remember when I did a branch on mixnet (it did change a lot in design since), I did two different integration: one that listen on different events and allows to pass the mixnet info on connection (and with key derived from the grandpa key), and a second one that just published in the dht.

I did like the dht branch a lot more but it was more experimental as the routing table of each node of the mixnet was also transmitted, allowing limitted number of mixnet connection, so a lot more complicated, making things a bit uncertain compared with the other branch.

Thinking about it now, with this different design (topology with all mixnet node connected to all mixnet node), using the dht sounds good to me (can sometime miss some update but that would not be a catastrophe).

So I don't really remember the exact reason why we went for info on the blockchain. Likely for simplicity but hard to say.

One usecase I find interesting with on chain info would be to register an different node than the validation node, but would work as well with dht design.

zdave commented:
> Everyone can put data into the DHT, not only authorities.

Yes, my point was more that it would no longer be as simple as "use the authority discovery service".

What are the main reasons you'd rather have peer IDs and addresses in the DHT? I can think of the following reasons:

- It would allow mixnodes to publish new addresses after registration.
- It would reduce the size of registration transactions and the amount of data in the blockchain state.

Neither of these seem particularly compelling to me though.

arkpar commented:
One other conern with DHT is security. The goal of the mixnet is to hide network activity. It has to be a self-contained protocol to have security guarantees that can be reasoned about. Once you have additional protocols, such as DHT involved, you need to be sure that these protocols won't compromise overall security. E.g. the attacker should not be able to correlate DHT traffic to mixnet traffic. 

It is just safer and simpler to use publicly available on-chain data instead.

bkchr commented:
> E.g. the attacker should not be able to correlate DHT traffic to mixnet traffic.

Hmm, yeah. Okay then let's keep the data in the on chain data.

zdave commented:
Been on holiday for 2 weeks, just getting back to this.

Re moving the registration stuff from an offchain worker to a separate call into the runtime: I can do this, but it will involve duplicating some of the stuff in `client/offchain`, or making it public. Currently I'm calling `sp_io::offchain::random_seed()`, `sp_io::offchain::is_validator()`, and `sp_io::offchain::network_state()` from the offchain worker. I'll have to copy across some of the `Externalities` implementation from `client/offchain`, or I guess pass this stuff in as arguments.

It's not clear to me why the offchain stuff is handled differently to the other execution extensions -- why is the `Externalities` implementation attached to `ExecutionContext` rather than `ExecutionExtensions`? This would be easier if the latter was the case.

zdave-parity committed:
Update Cargo.lock

zdave commented on client/rpc-api/src/mixnet/error.rs:
```diff
@@ -0,0 +1,48 @@
+/// Mixnet RPC error type.
+pub struct Error(pub sc_mixnet::Error);
+
+/// Base code for all mixnet errors.
+const BASE_ERROR: i32 = 7000;
```
Done.

zdave commented on primitives/mixnet/src/host_api.rs:
```diff
@@ -0,0 +1,48 @@
+
+/// Mixnet host functions.
+#[cfg(feature = "std")]
+pub type Functions = (self::mixnet::HostFunctions,);
```
This is gone now.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,776 @@
+			Weight::zero() // TODO
+		}
+
+		fn offchain_worker(block_number: T::BlockNumber) {
+			// If the local node is not running as a validator, never try to register as a
+			// mixnode...
+			if sp_io::offchain::is_validator() {
+				if let Err(err) = Self::maybe_register_local_node(block_number) {
+					log::debug!(
+						target: LOG_TARGET,
+						"Mixnet registration at block {block_number:?}: {err}"
+					);
+				}
+			}
+		}
```
Done.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3277875

zdave commented:
I've reworked things to get rid of the offchain worker. The mixnet service now just calls the `MixnetApi::maybe_register` runtime function every block instead, which does basically the same thing.

arkpar commented:
@bkchr Could you please take another look?

zdave-parity committed:
Fix mixnode registration

zdave-parity committed:
Update spec URL

bkchr commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,675 @@
+	}
+
+	/// `session_index` must be the index of the current session. `authority_index` is the
+	/// authority index in the _next_ session. Should only be called during the register phase
+	/// (`DisconnectFromPrev`).
```
I find it really really confusing that you everywhere talk about "register phase" and then the name is `DisconnectFromPrev`.

zdave replied:
Done. Hopefully less confusing now.

bkchr commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,675 @@
+) -> u64 {
+	let block_number: u64 = block_number.unique_saturated_into();
+	let mut data = [0; 8 + KX_PUBLIC_SIZE];
+	*array_mut_ref![data, 0, 8] = block_number.to_le_bytes();
+	*array_mut_ref![data, 8, KX_PUBLIC_SIZE] = *kx_public;
```
```suggestion
	data[..8].copy_from_slice(block_number.to_le_bytes());
	data[8..].copy_from_slice(&*kx_public);
```

No need to import this extra crate.

zdave replied:
Done.

bkchr commented on frame/mixnet/Cargo.toml:
```diff
@@ -0,0 +1,49 @@
+targets = ["x86_64-unknown-linux-gnu"]
+
+[dependencies]
+arrayref = "0.3.6"
```
```suggestion
```

zdave replied:
Done.

bkchr commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,675 @@
+		/// transactions and the longevity of these transactions.
+		type NextSessionRotation: EstimateNextSessionRotation<BlockNumberFor<Self>>;
+
+		/// Length of the `CoverToCurrent` phase of each session, in blocks.
+		#[pallet::constant]
+		type NumCoverToCurrentBlocks: Get<BlockNumberFor<Self>>;
+
+		/// Length of the `RequestsToCurrent` phase of each session, in blocks.
+		#[pallet::constant]
+		type NumRequestsToCurrentBlocks: Get<BlockNumberFor<Self>>;
+
+		/// Length of the `CoverToPrev` phase of each session, in blocks.
+		#[pallet::constant]
+		type NumCoverToPrevBlocks: Get<BlockNumberFor<Self>>;
```
There should be some comment on the order of the phases.

zdave replied:
Done.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,675 @@
+	/// except, after `true` is returned, this should not be called for a few blocks, to give the
+	/// submitted extrinsic a chance to get included.
+	pub fn maybe_register(session_index: SessionIndex, mixnode: Mixnode) -> bool {
+		let current_session_index = T::ValidatorSet::session_index();
+		if session_index != current_session_index {
```
It's just to avoid needing two calls into the runtime every block (one to check the session index and one to register). I can drop it if you'd prefer the two calls.

zdave commented on client/mixnet/src/sync_with_runtime.rs:
```diff
@@ -0,0 +1,208 @@
+	if let Some(peer_id) = from_core_peer_id(&mixnode.peer_id) {
+		fixup_external_addresses(&mut external_addresses, &peer_id);
+	} else {
+		error!(target: LOG_TARGET,
+			"Failed to convert mixnet peer ID {:x?} to libp2p peer ID",
+			mixnode.peer_id);
```
Is there a reference somewhere for which log levels should be used when? For this mixnet stuff I have so far been using error for things which should only happen if some node misbehaves, warn for things which might happen even if all nodes behave (but isn't likely to), and debug for everything else.

zdave commented on client/mixnet/src/run.rs:
```diff
@@ -0,0 +1,381 @@
+						reply_manager.abandon(reply_context);
+					}
+				} else {
+					error!(target: LOG_TARGET,
```
It shouldn't happen. Perhaps this should be clearer.

zdave replied:
I've added "; this is a bug" onto the end of these log messages.

zdave commented on client/mixnet/src/run.rs:
```diff
@@ -0,0 +1,381 @@
+		None
+	};
+	let mut network_events = network.event_stream("mixnet").fuse();
+	let mut next_forward_packet_delay = MaybeInfDelay::new(None);
+	let mut next_authored_packet_delay = MaybeInfDelay::new(None);
```
Why?

zdave replied:
Sorry I see your comment above now...

zdave commented on client/mixnet/src/request.rs:
```diff
@@ -0,0 +1,112 @@
+	}
+}
+
+pub const SUBMIT_EXTRINSIC: u8 = 1;
```
This is what I had originally and @arkpar asked me to change it. I really am not bothered either way...

zdave commented on client/mixnet/src/api.rs:
```diff
@@ -0,0 +1,61 @@
+		(Self { request_sender }, ApiBackend { request_receiver })
+	}
+
+	/// Submit an extrinsic via the mixnet. Returns a [`Future`] which returns another [`Future`].
+	/// The second [`Future`] does not borrow `self`.
```
The last sentence is the whole reason for doing things this way. Perhaps I should elaborate in the comment. The first `Future` that is returned borrows `self` -- it waits for space to send a message to the mixnet service. The second `Future` is standalone -- it just waits on the oneshot channel. The point of this is to allow multiple concurrent RPC requests without creating multiple channels to the mixnet service. The calling code needs to be structured appropriately for this to work, hence the mention in the doc comment.

zdave replied:
Doc is hopefully clearer now.

zdave commented on client/mixnet/src/maybe_inf_delay.rs:
```diff
@@ -0,0 +1,111 @@
+/// Unlike [`Delay`], implements [`FusedFuture`], with [`is_terminated`](Self::is_terminated)
+/// returning `true` when the delay is infinite. As with [`Delay`], once [`poll`](Self::poll)
+/// returns [`Poll::Ready`], it will continue to do so until [`reset`](Self::reset) is called.
+pub struct MaybeInfDelay(Inner);
```
Yeah true. This was originally needed for the mixnet libp2p behaviour, but that's gone now and the `reset` behaviour is not needed for `sc_mixnet::run`. I used it there because it's likely a bit more efficient than the alternative (I believe resetting a `Delay` is a bit more efficient than discarding and recreating), but I agree that it's not worth all this extra code.

zdave replied:
Removed.

zdave replied:
Hmm actually scratch that. `Fuse<Delay>` does not have quite the right behaviour. Unlike `MaybeInfDelay`, it will only return `Ready` once when the delay elapses; after that, it will return `Pending` again. This does not mesh well with the way these delays are reset. The mixnet crate sets flags when the delays need to change. If there are say two messages in a row with the same timestamp, it is possible that the mixnet crate will not set the delay-needs-to-change flag between them. This is fine with the current code as the delay will trigger again. This would _not_ work properly with `Fuse<Delay>`.

zdave committed:
Drop ValidatorSet from pallet_mixnet::Config

Not necessary; the mixnet pallet can track the session index itself.

zdave committed:
Move session key types from pallet_mixnet to sp_mixnet

To avoid node depending on pallet_mixnet.

zdave commented on bin/node/cli/Cargo.toml:
```diff
@@ -94,6 +95,7 @@ pallet-assets = { version = "4.0.0-dev", path = "../../../frame/assets/" }
 pallet-asset-conversion-tx-payment = { version = "4.0.0-dev", path = "../../../frame/transaction-payment/asset-conversion-tx-payment" }
 pallet-asset-tx-payment = { version = "4.0.0-dev", path = "../../../frame/transaction-payment/asset-tx-payment" }
 pallet-im-online = { version = "4.0.0-dev", default-features = false, path = "../../../frame/im-online" }
+pallet-mixnet = { version = "0.1.0-dev", path = "../../../frame/mixnet" }
```
The session key type from this crate is needed by `chain_spec.rs`. I've moved the type to `sp-mixnet` as it seems most other session key types are defined in primitives.

zdave commented on client/cli/src/params/mixnet_params.rs:
```diff
@@ -0,0 +1,67 @@
+// This file is part of Substrate.
```
Hmm I'm actually thinking the kitchen-sink node would be the right place for this? It makes assumptions that are not made by `sc-mixnet`. For example, `sc-mixnet` does not assume that non-authorities cannot be mixnodes (`sc-mixnet` doesn't care whether nodes are authorities or not). Similarly, `sc-mixnet` doesn't really know anything about RPC (so the comments about RPC would be out of place).

zdave committed:
Drop years from copyright headers

zdave commented on client/cli/src/params/mixnet_params.rs:
```diff
@@ -0,0 +1,67 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/api.rs:
```diff
@@ -0,0 +1,61 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave committed:
Improve Api::submit_extrinsic doc

zdave commented on client/mixnet/src/config.rs:
```diff
@@ -0,0 +1,88 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/error.rs:
```diff
@@ -0,0 +1,56 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,44 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/maybe_inf_delay.rs:
```diff
@@ -0,0 +1,111 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/packet_dispatcher.rs:
```diff
@@ -0,0 +1,198 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/peer_id.rs:
```diff
@@ -0,0 +1,41 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/peer_id.rs:
```diff
@@ -0,0 +1,41 @@
+	public.try_into_ed25519().ok().map(|public| public.to_bytes())
+}
+
+/// Convert a mixnet core [`PeerId`](CorePeerId) into a libp2p [`PeerId`]. This will succeed only
+/// if `peer_id` represents a point on the Ed25519 curve. Returns `None` on failure.
```
Done.

zdave committed:
Doc formatting

zdave committed:
Some doc comments

zdave commented on client/mixnet/src/request.rs:
```diff
@@ -0,0 +1,112 @@
+	delay.to_duration(config.mean_extrinsic_delay)
+}
+
+pub enum Request {
```
Added some docs here.

zdave committed:
Clarify some log messages

zdave commented on client/mixnet/src/sync_with_runtime.rs:
```diff
@@ -0,0 +1,208 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,675 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave committed:
Avoid arrayref dependency

zdave committed:
Improve mixnet pallet comments

In particular, avoid referring to "the register phase". Just refer to it
as the DisconnectFromPrev phase, which is the name of the enum variant.

zdave commented on primitives/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,24 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on primitives/mixnet/src/runtime_api.rs:
```diff
@@ -0,0 +1,42 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave committed:
Improve maybe_register doc comment

zdave commented on primitives/mixnet/src/types.rs:
```diff
@@ -0,0 +1,93 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022-2023 Parity Technologies (UK) Ltd.
```
Done.

zdave commented on client/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,41 @@
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate mixnet service.
```
Linked to the spec.

zdave commented on frame/mixnet/src/lib.rs:
```diff
@@ -0,0 +1,776 @@
+	/// very first session the authority indices for the previous mixnode set are made up.
+	#[pallet::storage]
+	pub(crate) type OddSessionMixnodes<T> =
+		StorageMap<_, Identity, AuthorityIndex, BoundedOpaqueMixnodeFor<T>>;
```
Done, thanks. `StorageDoubleMap` has made things a lot simpler.

zdave committed:
Simplify mixnode storage in mixnet pallet

Note that registrations are now accepted in all phases, and the previous
mixnode set is kept for the whole session.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3426352

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3426351
